### PR TITLE
[XLA] [GPU] Support int8 in BufferComparator

### PR DESCRIPTION
**Pull request by yongfeng-nv**
…llision.

@cheshire I just changed the version number to 4.2 and replaced tabs with spaces.  Could you try it again?

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/buffer_comparator.cc [hunk 1]
```diff
@@ -405,11 +619,13 @@ StatusOr<bool> HostCompare(se::Stream* stream, se::DeviceMemoryBase lhs,
 
   const auto canonicalize = [](ComparisonType a) -> ComparisonType {
     if (std::is_same<ElementType, Eigen::half>::value && a) {
-      constexpr ComparisonType kMaxFp16Value = 65505.;
+      constexpr ComparisonType kMaxFp16Value =
+          std::is_same<ElementType, Eigen::half>::value ? 65505. : 0;
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/buffer_comparator.cc]: Aren't you already inside the branch which guarantees you that `ElementType` is Eigen::half?

File: tensorflow/compiler/xla/service/gpu/buffer_comparator.cc [hunk 2]
```diff
@@ -405,11 +619,13 @@ StatusOr<bool> HostCompare(se::Stream* stream, se::DeviceMemoryBase lhs,
...
       if (std::isnan(a)) {
         return a;
       }
-      return std::max(-kMaxFp16Value, std::min(a, kMaxFp16Value));
+      return std::max(static_cast<ComparisonType>(-kMaxFp16Value),
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/buffer_comparator.cc]: Why the change?

**Review by cheshire — CHANGES_REQUESTED** [review]
Two minor comments, otherwise LGTM

File: tensorflow/compiler/xla/service/gpu/buffer_comparator.cc (hunk 1 above)
Comment (yongfeng-nv) [reply tensorflow/compiler/xla/service/gpu/buffer_comparator.cc]: I agree, but compiler still errors about overflow.  I added the following comments on this in the previous PR:

This change is needed, because 65505 can't fit in int8. Refactoring this lambda to a set of template/overloaded helper functions is another solution.

Comment (yongfeng-nv) [reply tensorflow/compiler/xla/service/gpu/buffer_comparator.cc]: My bad.  The error was caused by a wrong comparison type.  With comparison type being float, this lambda doesn't have to change.  Reverting the changes.

File: tensorflow/compiler/xla/service/gpu/buffer_comparator.cc (hunk 2 above)
Comment (yongfeng-nv) [reply tensorflow/compiler/xla/service/gpu/buffer_comparator.cc]: This is not necessary either.

**Review by cheshire — DISMISSED** [review]

**Comment by cheshire** [comment]
@yongfeng-nv  Could you take a look at a conflict and rebase? I think an unrelated change erroneously addded some whitespace to buffer_comparator, it looks like the other changes could be discarded.

**Comment by yongfeng-nv** [comment]
> @yongfeng-nv Could you take a look at a conflict and rebase? I think an unrelated change erroneously addded some whitespace to buffer_comparator, it looks like the other changes could be discarded.

Just merged and updated the PR.

**Review by cheshire — APPROVED** [review]
